File: app/src/main/java/com/orego/battlecrane/bc/std/scenario/skirmish/BStandardSkirmishScenario.java
package com.orego.battlecrane.bc.std.scenario.skirmish;

import com.orego.battlecrane.bc.api.context.BGameContext;
import com.orego.battlecrane.bc.api.context.controller.map.BMapController;
import com.orego.battlecrane.bc.api.model.player.BPlayer;
import com.orego.battlecrane.bc.api.model.contract.BUnit;
import com.orego.battlecrane.bc.api.scenario.BGameScenario;
import com.orego.battlecrane.bc.std.location.grass.field.empty.BEmptyField;
import com.orego.battlecrane.bc.std.race.human.unit.building.implementation.BHumanHeadquarters;
import com.orego.battlecrane.bc.std.race.human.unit.building.implementation.BHumanWall;
import com.orego.battlecrane.bc.std.race.human.scenario.skirmish.adjutant.BHumanAdjutant;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import static com.orego.battlecrane.bc.api.context.controller.map.BMapController.MAP_SIZE;

public final class BStandardSkirmishScenario implements BGameScenario {

    @NotNull
    @Override
    public final List<BPlayer> initPlayerList(final @NotNull BGameContext context) {
        final List<BPlayer> playerList = new ArrayList<>();
        final BPlayer redPlayer = new BPlayer(context, new BHumanAdjutant.Builder());
        final BPlayer bluePlayer = new BPlayer(context, new BHumanAdjutant.Builder());
        //Set enemies:
        redPlayer.addEnemy(bluePlayer);
        bluePlayer.addEnemy(redPlayer);
        //Add in player list:
        playerList.add(bluePlayer);
        playerList.add(redPlayer);
        return playerList;
    }

    @NotNull
    @Override
    public final BPlayer getStartPlayerPosition(final List<BPlayer> players) {
        return players.get(new Random().nextInt(1));
    }

    @Override
    public final void initMap(
            @NotNull final BMapController.BMapHolder mapHolder,
            @NotNull final BGameContext context
    ) {
        //Fill startTurn player stackPosition:
        final List<BPlayer> players = context.getPlayerManager().getPlayers();
        if (players.size() == 2) {
            //Get players:
            final BPlayer bluePlayer = players.get(0);
            final BPlayer redPlayer = players.get(1);
            //Put headquarters on the map:
            mapHolder.bindUnitTo(new BHumanHeadquarters(context, bluePlayer), 14, 14);
            mapHolder.bindUnitTo(new BHumanHeadquarters(context, redPlayer), 0, 0);
            //Put walls on the map
            for (int j = 0; j < 5; j++) {
                mapHolder.bindUnitTo(new BHumanWall(context, redPlayer), j, 4);
                mapHolder.bindUnitTo(new BHumanWall(context, redPlayer), 4, j);
                mapHolder.bindUnitTo(new BHumanWall(context, bluePlayer), 15 - j, 11);
                mapHolder.bindUnitTo(new BHumanWall(context, bluePlayer), 11, 15 - j);
            }
        } else {
            throw new IllegalArgumentException("Standard skirmish scenario supports two players!");
        }
        //Fill rest points:
        for (int x = 0; x < MAP_SIZE; x++) {
            for (int y = 0; y < MAP_SIZE; y++) {
                if (!mapHolder.hasBoundUnit(x, y)) {
                    final BUnit emptyField = new BEmptyField(context);
                    mapHolder.bindUnitTo(emptyField, x, y);
                }
            }
        }
    }
}

File: app/src/main/java/com/orego/battlecrane/bc/std/scenario/skirmish/BStandardSkirmishScenario.kt
package com.orego.battlecrane.bc.std.scenario.skirmish

import com.orego.battlecrane.bc.api.context.BGameContext
import com.orego.battlecrane.bc.api.context.controller.map.BMapController
import com.orego.battlecrane.bc.api.context.storage.heap.implementation.BPlayerHeap
import com.orego.battlecrane.bc.api.model.adjutant.BAdjutant
import com.orego.battlecrane.bc.api.model.entity.main.unit.BUnit
import com.orego.battlecrane.bc.api.model.player.BPlayer
import com.orego.battlecrane.bc.api.scenario.BGameScenario
import com.orego.battlecrane.bc.std.location.grass.field.implementations.empty.BEmptyGrassField
import com.orego.battlecrane.bc.std.race.human.unit.building.implementation.headquarters.BHumanHeadquarters
import com.orego.battlecrane.bc.std.race.human.unit.building.implementation.wall.BHumanWall
import com.orego.battlecrane.bc.std.scenario.skirmish.model.player.BStandardSkirmishPlayerBuilder
import com.orego.battlecrane.bc.std.scenario.skirmish.model.race.human.adjutant.BStandardSkirmishHumanAdjutantBuilder
import java.util.*

class BStandardSkirmishScenario : BGameScenario {

    override val startTurnPlayerPosition = Random().nextInt(1)

    override fun getPlayers(context: BGameContext): List<BPlayer> {
        val playerList = mutableListOf<BPlayer>()
        val builder = BStandardSkirmishPlayerBuilder()
        val redPlayer = builder.build(context)
        val bluePlayer = builder.build(context)

        //Set enemies:
        redPlayer.addEnemy(bluePlayer.playerId)
        bluePlayer.addEnemy(redPlayer.playerId)

        //Add in player list:
        playerList.add(bluePlayer)
        playerList.add(redPlayer)
        return playerList
    }

    override fun getAdjutants(context: BGameContext): List<BAdjutant> {
        val adjutantList = mutableListOf<BAdjutant>()
        val heap = context.storage.getHeap(BPlayerHeap::class.java)
        val players = heap.getObjectList()
        val builder = BStandardSkirmishHumanAdjutantBuilder()
        if (players.size == 2) {
            adjutantList.add(builder.build(context, players[0].playerId))
            adjutantList.add(builder.build(context, players[1].playerId))
            return adjutantList
        } else {
            throw IllegalArgumentException("Standard skirmish gameScenario supports two players!")
        }
    }

    override fun getUnits(context: BGameContext) =
        mutableListOf<BUnit>()
            .addBuildings(context)
            .addEmptyFields(context)

    private fun MutableList<BUnit>.addBuildings(context: BGameContext): MutableList<BUnit> {
        val heap = context.storage.getHeap(BPlayerHeap::class.java)
        val players = heap.getObjectList()
        if (players.size == 2) {

            //Get players:
            val bluePlayerId = players[0].playerId
            val redPlayerId = players[1].playerId

            //Put headquarters on the mapConstraintLayout:
            this.add(
                BHumanHeadquarters(
                    context,
                    bluePlayerId,
                    14,
                    14
                )
            )
            this.add(
                BHumanHeadquarters(
                    context,
                    redPlayerId,
                    0,
                    0
                )
            )

            //Put walls on the mapConstraintLayout
            for (j in 0..4) {
                this.add(
                    BHumanWall(
                        context,
                        redPlayerId,
                        j,
                        4
                    )
                )
                this.add(
                    BHumanWall(
                        context,
                        redPlayerId,
                        4,
                        j
                    )
                )
                this.add(
                    BHumanWall(
                        context,
                        bluePlayerId,
                        15 - j,
                        11
                    )
                )
                this.add(
                    BHumanWall(
                        context,
                        bluePlayerId,
                        11,
                        15 - j
                    )
                )
            }
            return this
        } else {
            throw IllegalArgumentException("Standard skirmish gameScenario supports two players!")
        }
    }

    private fun MutableList<BUnit>.addEmptyFields(context: BGameContext): MutableList<BUnit> {
        val matrix = Array(BMapController.MAP_SIZE) {
            Array(BMapController.MAP_SIZE) {
                BMapController.NOT_INITIALIZED_UNIT_ID
            }
        }
        //Check unit list:
        for (i in 0 until this.size) {
            val unit = this[i]
            val unitId = unit.unitId
            val startX = unit.x
            val startY = unit.y
            val endX = startX + unit.width
            val endY = startY + unit.height
            for (x in startX until endX) {
                for (y in startY until endY) {
                    matrix[x][y] = unitId
                }
            }
        }
        //Fill rest fileds:
        val emptyGrassBuilder = BEmptyGrassField.Builder()
        for (x in 0 until BMapController.MAP_SIZE) {
            for (y in 0 until BMapController.MAP_SIZE) {
                if (matrix[x][y] == BMapController.NOT_INITIALIZED_UNIT_ID) {
                    this.add(emptyGrassBuilder.build(context, BPlayer.NEUTRAL_PLAYER_ID, x, y))
                }
            }
        }
        return this
    }
}